src/relations/db.py

```diff
@@ -0,0 +1,414 @@
+│                  │            port │ 5432                                       │               │
+│                  │        standbys │ host=pgbouncer-k8s-operator-0.pgbouncer-k… │               │
+│                  │                 │ dbname=waltz port=6432                     │               │
+│                  │                 │ fallback_application_name=finos-waltz      │               │
+│                  │                 │ user=relation_id_3 password=ZFz5RnH7hyvTp… │               │
+│                  │           state │ master                                     │               │
+│                  │            user │ relation_id_3                              │               │
+│                  │         version │ 12                                         │               │
+└──────────────────┴─────────────────┴────────────────────────────────────────────┴───────────────┘
+
+"""
+
+import logging
+from typing import Iterable
+
+from charms.pgbouncer_operator.v0 import pgb
+from charms.postgresql_k8s.v0.postgresql import (
+    PostgreSQLCreateDatabaseError,
+    PostgreSQLCreateUserError,
+    PostgreSQLDeleteUserError,
+)
+from ops.charm import (
+    CharmBase,
+    RelationBrokenEvent,
+    RelationChangedEvent,
+    RelationDepartedEvent,
+    RelationJoinedEvent,
+)
+from ops.framework import Object
+from ops.model import (
+    ActiveStatus,
+    Application,
+    BlockedStatus,
+    MaintenanceStatus,
+    Relation,
+    Unit,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class DbProvides(Object):
+    """Defines functionality for the 'provides' side of the 'db' relation.
+
+    Hook events observed:
+        - relation-changed
+        - relation-departed
+        - relation-broken
+    """
+
+    def __init__(self, charm: CharmBase, admin: bool = False):
+        """Constructor for DbProvides object.
+
+        Args:
+            charm: the charm for which this relation is provided
+            admin: a boolean defining whether or not this relation has admin permissions, switching
+                between "db" and "db-admin" relations.
+        """
+        if admin:
+            self.relation_name = "db-admin"
+        else:
+            self.relation_name = "db"
+
+        super().__init__(charm, self.relation_name)
+
+        self.framework.observe(
+            charm.on[self.relation_name].relation_joined, self._on_relation_joined
+        )
+        self.framework.observe(
+            charm.on[self.relation_name].relation_changed, self._on_relation_changed
+        )
+        self.framework.observe(
+            charm.on[self.relation_name].relation_departed, self._on_relation_departed
+        )
+        self.framework.observe(
+            charm.on[self.relation_name].relation_broken, self._on_relation_broken
+        )
+
+        self.charm = charm
+        self.admin = admin
+
+    def _on_relation_joined(self, join_event: RelationJoinedEvent):
+        """Handle db-relation-joined event."""
```

Comment:
You could add a summary to the method description explaining what is done in this handler like you did in the relation changed handler.

Reply:
good idea!